File: apps/chess-coach/harness/src/main/kotlin/harness/Orchestrator.kt
package harness

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

@Serializable
data class UiPhrases(val thinking: List<String>, val bestMove: List<String>)

class Orchestrator(
    private val stateManager: StateManager,
    private val engineBridge: EngineBridge,
    private val llmClient: LlmClient,
    private val skillLoader: SkillLoader = SkillLoader()
) {
    suspend fun executeTurn(humanMove: String, difficulty: String = "intermediate"): TurnResult {
        System.err.println("--- Starting Turn ---")
        System.err.println("Human played: ${if (humanMove.isEmpty()) "(None - First Move)" else humanMove}")
        
        // 1. Read Current State
        val currentFen = stateManager.readFen()
        
        // Determine colors based on whose turn it is in the FEN
        val activeColor = currentFen.split(" ").getOrNull(1) ?: "w"
        val aiColor = if (activeColor == "w") "White" else "Black"
        
        // 2. Get Board Evaluation
        System.err.println("Evaluating position...")
        val evalResult = engineBridge.getEvaluation(currentFen, depth = 15)
        val evalString = if (evalResult.isMate) "Mate in ${evalResult.mateIn}" else String.format("%.2f", evalResult.cp / 100.0)
        System.err.println("Stockfish Eval: Move: ${evalResult.bestMove}, Eval: $evalString")
        
        // 3. Get Move from Maia (lc0)
        System.err.println("Asking Maia ($difficulty) for move...")
        val candidateMove = engineBridge.getMaiaMove(currentFen, difficulty)
        System.err.println("Maia Suggested Move: $candidateMove")
        
        // 4. Verify Candidate SAN
        val newFen = engineBridge.getFenAfterMove(currentFen, candidateMove) 
            ?: throw IllegalStateException("Maia generated an illegal move: $candidateMove")
        
        System.err.println("Move '$candidateMove' is LEGAL.")
        
        // 5. Commit Valid Move
        stateManager.writeFen(newFen)
        
        return TurnResult(
            fen = newFen,
            move = candidateMove,
            advice = ""
        )
    }

    suspend fun generateAdvice(humanMove: String, aiMove: String, currentFen: String): String {
        System.err.println("Generating coaching advice...")
        val evalResult = engineBridge.getEvaluation(currentFen, depth = 15)
        
        val activeColor = currentFen.split(" ").getOrNull(1) ?: "w"
        val aiColor = if (activeColor == "w") "Black" else "White"
        val humanColor = if (aiColor == "White") "Black" else "White"

        val adviceSystemPrompt = "Generate professional chess commentary in the specified language. Always use Standard Algebraic Notation (SAN) for moves (e.g., Nf3, e4). For Type=standard use 30–40 words. For Type=explanation, explain the best move briefly (≤50 words). Return exactly: Commentary, Predicted ELO, Verified Classification."
        
        // We approximate the CP delta since we don't have the exact previous evaluation stored
        val cpString = if (evalResult.isMate) "Mate in ${evalResult.mateIn}" else "${evalResult.cp}->${evalResult.cp} (Δ=0)"
        val safeHumanMove = if (humanMove.isBlank()) "e4" else humanMove

        val adviceUserPrompt = """LanguageL: English
LangCode: en
Type: standard
FEN: $currentFen
MoveSAN: $safeHumanMove
Side: $humanColor
Actor: human
Gender: neutral
Tag: Good
BestAlt: ${evalResult.bestMove}
CP: $cpString"""
        
        val rawAdvice = llmClient.prompt(adviceSystemPrompt, adviceUserPrompt, llmClient.commentaryModel, temperature = 0.7)
        System.err.println("Coach Advice generated.")
        
        // Extract just the commentary part if the model returns the full block
        val commentaryMatch = Regex("Commentary:\\s*(.*?)(?:\\n|Predicted ELO:|$)", RegexOption.DOT_MATCHES_ALL).find(rawAdvice)
        return commentaryMatch?.groupValues?.get(1)?.trim() ?: rawAdvice
    }

    suspend fun generateExplanation(currentFen: String): String {
        System.err.println("Generating blunder explanation...")
        val evalResult = engineBridge.getEvaluation(currentFen, depth = 15)
        
        val activeColor = currentFen.split(" ").getOrNull(1) ?: "w"
        val humanColor = if (activeColor == "w") "White" else "Black"

        val explainSystemPrompt = "You are a helpful chess coach. Explain briefly (under 40 words) why the last move resulting in this position is a blunder or mistake. Focus on the immediate tactical problem."
        
        val cpString = if (evalResult.isMate) "Mate in ${evalResult.mateIn}" else "${evalResult.cp}"

        val explainUserPrompt = """FEN: $currentFen
Side to move: $humanColor
Best alternative: ${evalResult.bestMove}
Current Eval: $cpString"""
        
        val rawExplanation = llmClient.prompt(explainSystemPrompt, explainUserPrompt, llmClient.commentaryModel, temperature = 0.7)
        System.err.println("Explanation generated.")
        
        return rawExplanation.trim()
    }

    suspend fun generateUiPhrases(): UiPhrases {
        return UiPhrases(
            thinking = listOf("Hmm...", "Let me think...", "Interesting position...", "Rats...", "What to do..."),
            bestMove = listOf("Great move!", "Excellent!", "I like that.", "Strong play.", "Well done.")
        )
    }
}

data class TurnResult(
    val fen: String,
    val move: String,
    val advice: String
)

File: apps/chess-coach/ui/src/engine/Orchestrator.ts
import { EngineBridge } from './EngineBridge';
import { LlmClient } from './LlmClient';

export type TurnResult = { fen: string; move: string; advice: string };
export type UiPhrases = { thinking: string[]; bestMove: string[] };

export class Orchestrator {
  private engineBridge = new EngineBridge();
  private llmClient = new LlmClient();
  private currentFen = 'rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1';

  resetGame(): string {
    this.currentFen = 'rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1';
    return this.currentFen;
  }

  async executeTurn(humanMoveSan: string, fenAfterHuman: string, difficulty: string): Promise<TurnResult> {
    console.log(`--- Starting Turn ---`);
    console.log(`Human played: ${humanMoveSan || '(None - First Move)'}`);
    
    this.currentFen = fenAfterHuman;

    console.log('Evaluating position...');
    const evalResult = await this.engineBridge.getEvaluation(this.currentFen, 15);
    const evalString = evalResult.isMate ? `Mate in ${evalResult.mateIn}` : (evalResult.cp / 100.0).toFixed(2);
    console.log(`Stockfish Eval: Move: ${evalResult.bestMove}, Eval: ${evalString}`);

    console.log(`Asking AI (${difficulty}) for move...`);
    const candidateMove = await this.engineBridge.getAiMove(this.currentFen, difficulty);
    console.log(`AI Suggested Move: ${candidateMove}`);

    const newFen = await this.engineBridge.getFenAfterMove(this.currentFen, candidateMove);
    if (!newFen) {
      throw new Error(`AI generated an illegal move: ${candidateMove}`);
    }

    console.log(`Move '${candidateMove}' is LEGAL.`);
    this.currentFen = newFen;

    return {
      fen: newFen,
      move: candidateMove,
      advice: ''
    };
  }

  async *generateAdviceStream(humanMove: string, _aiMove: string, currentFen: string): AsyncGenerator<string, void, unknown> {
    console.log('Generating coaching advice stream...');
    const evalResult = await this.engineBridge.getEvaluation(currentFen, 15);
    
    const parts = currentFen.split(' ');
    const activeColor = parts[1] || 'w';
    const aiColor = activeColor === 'w' ? 'Black' : 'White';
    const humanColor = aiColor === 'White' ? 'Black' : 'White';

    const adviceSystemPrompt = "Generate professional chess commentary in the specified language. Always use Standard Algebraic Notation (SAN) for moves (e.g., Nf3, e4). For Type=standard use 30–40 words. For Type=explanation, explain the best move briefly (≤50 words). Return exactly: Commentary, Predicted ELO, Verified Classification.";
    
    const cpString = evalResult.isMate ? `Mate in ${evalResult.mateIn}` : `${evalResult.cp}->${evalResult.cp} (Δ=0)`;
    const safeHumanMove = humanMove.trim() === '' ? 'e4' : humanMove;

    const adviceUserPrompt = `LanguageL: English
LangCode: en
Type: standard
FEN: ${currentFen}
MoveSAN: ${safeHumanMove}
Side: ${humanColor}
Actor: human
Gender: neutral
Tag: Good
BestAlt: ${evalResult.bestMove}
CP: ${cpString}`;

    yield* this.llmClient.promptStream(adviceSystemPrompt, adviceUserPrompt, 0.7, 150);
  }

  async *generateExplanationStream(fenBefore: string, fenAfter: string, isBlunder: boolean, moveSan: string): AsyncGenerator<string, void, unknown> {
    console.log(`Generating explanation stream (isBlunder=${isBlunder}, move=${moveSan})...`);
    const evalBefore = await this.engineBridge.getEvaluation(fenBefore, 15);
    const evalAfter = await this.engineBridge.getEvaluation(fenAfter, 15);
    
    const partsAfter = fenAfter.split(' ');
    const activeColorAfter = partsAfter[1] || 'w';
    const moveColor = activeColorAfter === 'w' ? 'Black' : 'White';
    const opponentColor = activeColorAfter === 'w' ? 'White' : 'Black';

    const explainSystemPrompt = isBlunder
        ? `You are a chess expert. The player playing ${moveColor} just played ${moveSan}, which is a blunder. Explain in exactly ONE short sentence why ${moveSan} is a blunder. Do not provide any formatting, greetings, or extra text.`
        : `You are a chess expert. The player playing ${moveColor} just played ${moveSan}, which is an excellent move. Explain in exactly ONE short sentence why ${moveSan} is strong. Do not provide any formatting, greetings, or extra text.`;
    
    const cpBefore = evalBefore.isMate ? `Mate in ${evalBefore.mateIn}` : `${evalBefore.cp}`;
    const cpAfter = evalAfter.isMate ? `Mate in ${evalAfter.mateIn}` : `${evalAfter.cp}`;

    const explainUserPrompt = isBlunder
        ? `Before FEN: ${fenBefore}\nAfter FEN: ${fenAfter}\nBlunder played: ${moveSan} (by ${moveColor})\nEval Before: ${cpBefore}\nEval After: ${cpAfter}\nBest continuation for ${opponentColor}: ${evalAfter.pv}`
        : `Before FEN: ${fenBefore}\nAfter FEN: ${fenAfter}\nGreat move played: ${moveSan} (by ${moveColor})\nEval Before: ${cpBefore}\nEval After: ${cpAfter}\nBest continuation: ${evalAfter.pv}`;
    
    yield* this.llmClient.promptStream(explainSystemPrompt, explainUserPrompt, 0.7, 80);
  }

  async generateUiPhrases(): Promise<UiPhrases> {
    try {
      console.log("Warming up LLM into VRAM...");
      await this.llmClient.prompt("System", "Ping", 0.7, 1);
      console.log("LLM Warmup complete.");
    } catch (e: any) {
      console.error(`LLM Warmup failed: ${e.message}`);
    }

    return {
      thinking: ["Hmm...", "Let me think...", "Interesting position...", "Rats...", "What to do..."],
      bestMove: ["Great move!", "Excellent!", "I like that.", "Strong play.", "Well done."]
    };
  }
}
